File: src/cli/python/target.py
import platform
from pathlib import Path
from typing import List
import os


class Target:
    def __init__(self, arch="unknown", os="unknown", env=None):
        self.arch = arch
        self.os = os
        self.env = env if env else Target.get_default_env(os)

    @staticmethod
    def from_host() -> 'Target':
        target = Target()

        machine = platform.machine().lower()
        if machine == "x86_64" or machine == "amd64":
            target.arch = "x86_64"
        elif machine == "aarch64" or machine == "arm64":
            target.arch = "aarch64"

        system = platform.system()
        if system == "Windows":
            target.os = "windows"
        elif system == "Linux":
            target.os = "linux"
        elif system == "Darwin":
            target.os = "macos"

        target.env = Target.get_default_env(target.os)
        return target

    @staticmethod
    def from_triple(triple) -> 'Target':
        components = triple.split("-")
        arch = components[0]
        os = components[1]
        env = components[2] if len(components) > 2 else Target.get_default_env(os)
        return Target(arch, os, env)

    @staticmethod
    def list_available() -> List['Target']:
        return [
            Target("x86_64", "windows", "msvc"),
            Target("x86_64", "windows", "gnu"),
            Target("x86_64", "linux", "gnu"),
            Target("x86_64", "macos"),
            Target("x86_64", "android"),
            Target("aarch64", "windows", "msvc"),
            Target("aarch64", "linux", "gnu"),
            Target("aarch64", "macos"),
            Target("aarch64", "android"),
            Target("aarch64", "ios")
        ]

    @property
    def triple(self):
        if self.env:
            return f"{self.arch}-{self.os}-{self.env}"
        else:
            return f"{self.arch}-{self.os}"

    @property
    def toolchain_file_path(self):
        return get_toolchains_dir() / (str(self) + ".json")

    @staticmethod
    def get_default_env(os):
        if os == "windows":
            return "msvc"
        elif os == "linux":
            return "gnu"
        else:
            return None

    def __str__(self):
        return self.triple

    def __eq__(self, other):
        return self.arch == other.arch and self.os == other.os and self.env == other.env
    
    def requires_assembler(self):
        return self not in (
            Target("x86_64", "windows", "msvc"),
            Target("x86_64", "windows", "gnu"),
            Target("x86_64", "linux", "gnu"),
            Target("aarch64", "macos"),
        )


def get_toolchains_dir():
    # if platform.system() == "Windows":
    #     toolchains_dir = Path(os.getenv("LOCALAPPDATA")) / "Programs" / "Banjo" / "toolchains"
    # else:
    #     toolchains_dir = Path.home() / "banjo" / "toolchains"

    toolchains_dir = Path.home() / ".banjo" / "toolchains"
    toolchains_dir.mkdir(parents=True, exist_ok=True)
    return toolchains_dir

File: src/cli/target.cpp
#include "target.hpp"

#include "banjo/utils/platform.hpp"

#include <utility>

namespace banjo {
namespace cli {

Target::Target() : arch("unknown"), os("unknown") {}

Target::Target(std::string arch, std::string os) : arch(std::move(arch)), os(std::move(os)), env() {}

Target::Target(std::string arch, std::string os, std::optional<std::string> env)
  : arch(std::move(arch)),
    os(std::move(os)),
    env(std::move(env)) {}

Target Target::host() {
    std::string arch;

#if ARCH_X86_64
    arch = "x86_64";
#elif ARCH_AARCH64
    arch = "aarch64";
#else
    arch = "unknown";
#endif

    std::string os;

#if OS_WINDOWS
    os = "windows";
#elif OS_LINUX
    os = "linux";
#elif OS_MACOS
    os = "macos";
#else
    os = "unknown";
#endif

    std::optional<std::string> env = get_default_env(os);

    return Target(arch, os, env);
}

std::string Target::to_string() const {
    if (env) {
        return arch + "-" + os + "-" + *env;
    } else {
        return arch + "-" + os;
    }
}

bool Target::is_executable_on_host() {
    Target host_target = host();

    if (*this == host_target) {
        return true;
    }

    if (host_target.os == "windows" && os == "windows" && env == "gnu") {
        return true;
    }

    return false;
}

std::vector<Target> Target::list_available() {
    return {
        Target("x86_64", "windows", "msvc"),
        Target("x86_64", "windows", "gnu"),
        Target("x86_64", "linux", "gnu"),
        Target("aarch64", "linux", "gnu"),
        Target("aarch64", "macos"),
        Target("wasm", "unknown"),
        Target("wasm", "emscripten"),
    };
}

std::optional<std::string> Target::get_default_env(const std::string &os) {
    if (os == "windows") {
        return "msvc";
    } else if (os == "linux") {
        return "gnu";
    } else {
        return {};
    }
}

} // namespace cli
} // namespace banjo
